README.md
<div id="top"></div>

> HTTPFY curently in beta so you may see problems. Please open a Issue on GitHub and report them!

<br />
<div align="center">
  <a href="https://github.com/devxprite/httpfy">
    <img src="static/httpFy_logo.png" alt="Logo" width="280">
  </a>

  <p align="center">
    A Incredible fast and Powerful HTTP toolkit
    <br>
    <br>
    <img alt="License" src="https://img.shields.io/npm/l/httpfy">
    <img alt="GitHub Repo stars" src="https://img.shields.io/github/stars/devxprite/httpfy">
    <img alt="version" src="https://img.shields.io/npm/v/httpfy">
    <br>
    <img alt="downloads" src="https://img.shields.io/npm/dw/httpfy">
    <img alt="Libraries.io SourceRank" src="https://img.shields.io/librariesio/sourcerank/npm/httpfy">
    <br />
    <a href="https://github.com/devxprite/httpfy/issues">Report Bug</a>
    ·
    <a href="https://github.com/devxprite/httpfy/issues">Request Feature</a>
  </p>
</div>

<!-- GETTING STARTED -->

## Getting Started

  ![Basic](static/preview.gif)

### Prerequisites


You need **NodeJs** 10 or later to run this tool. You can install NodeJs :

- On Linux
  ```sh
  sudo apt-get install nodejs
  ```
- On Termux
  ```sh
  pkg install nodejs-lts
  ```

## Installation

### Using npm

```bash
npm install -g -s httpfy
```
### Install locally

```bash
git clone https://github.com/devXprite/httpfy.git
cd httpfy
npm install
node index.js -h
```

### Uninstall

```bash
npm remove httpfy -g
```

## Usage


```bash
httpfy -f <filename containing urls>
```

### Options

| Short  | Options             | Type      | Description                                                   |
| ------ | ------------------- | --------- | ------------------------------------------------------------- |
| -v     | --version           | probe     | output the current version                                    |
| -f     | --file              | string    | input file containing list of URLs                            |
| -sc    | --status-code       | probe     | display response status-code                                  |
| -cl    | --content-length    | probe     | display response content-length                               |
| -ct    | --content-type      | probe     | display response content-type                                 |
| -rt    | --response-time     | probe     | display response time                                         |
| -lc    | --line-count        | probe     | display response body line count                              |
| -wc    | --word-count        | probe     | display response body word count                              |
| -ws    | --web-serve         | probe     | display web server name                                       |
| -rl    | --redirect-location | probe     | display redirect location                                     |
| -m     | --method            | probe     | display http request method                                   |
| -nc    | --no-color          | probe     | disable colors in cli output                                  |
| -ttl   | --title             | probe     | display page title                                            |
| -fl    | --failed            | probe     | display failed request's                                      |
| -nr    | --no-redirect       | probe     | don't follow redirects                                        |
| -mr    | --max-redirect      | number    | maximum redirects to follow                                   |
| -t     | --threads           | number    | maximum cocurrent requests send (default: 100)                |
| -i     | --interval          | number    | interval between each thread in seconds                       |
| -c     | --cookie            | string    | send cookies (--cookie 'login=Yes')                           |
| -H     | --header            | string    | custom http headers to send (--header 'X-MyHeader: XYZ')      |
| -x     | --request-methods   | string    | set request methods use 'all' to probe all HTTP methods       |
| -pc    | --protocol          | string    | set request Protocol (default: "https")                       |
| -path  | --request-path      | string    | path or list of paths  (-path admin,login)                    |
| -param | --request-param     | string    | set request parameters  (-param id=5)                         |
| -ua    | --user-agent        | string    | set custom useragent                                          |
| -time  | --timeout           | number    | set request timeout in seconds                                |
| -mc    | --match-code        | numbers[] | match response with specified status code  (-mc 200,404)      |
| -ml    | --match-length      | numbers[] | match response with specified content length  (-ml 800,900)   |
| -mlc   | --match-line-count  | numbers[] | match response body with specified line count  (-mlc 100,102) |
| -ms    | --match-string      | strings[] | match response with specified strings  (-ms admin)            |
| -o     | --output-file       | string    | save results in a single file  (-o result.txt)                |
| -of    | --output-folder     | string    | save response of url in multiple files  (-of result)          |
| -h     | --help              | probe     | display help for command                                      |

## Examples

### Basic

- ```bash
  httpfy -f urls.txt -sc -ttl -fl
  ```
  ![Basic](static/example_basic.gif)
### Try all Http Methods

- ```bash
  httpfy -f urls.txt -sc -m -x all
  ```
  ![All](static/example_all.gif)

### File/Path Bruteforce

- ```bash
  httpfy -f urls.txt -path admin,login -sc
  ```
  ![Brute](static/example_brute.gif)


<!-- GitPod -->

## Run with Gitpod

Click this button to run your project on Gitpod which comes with pre-configured environment you need.

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/#/https://github.com/devXprite/httpfy)

<!-- CONTRIBUTING -->
## Contributing

Contributions are what make the open source community such an amazing place to learn, inspire, and create. Any contributions you make are **greatly appreciated**.

If you have a suggestion that would make this better, please fork the repo and create a pull request. You can also simply open an issue with the tag "enhancement".
Don't forget to give the project a star! Thanks again!

1. Fork the Project
2. Create your Branch 
3. Commit your Changes 
4. Push to the Branch 
5. Open a Pull Request
 
## Supporters

### Stargazers:
[![Stargazers repo roster for @devXprite/httpfy](https://reporoster.com/stars/dark/devXprite/httpfy)](https://github.com/devXprite/httpfy/stargazers)  

### Forkers
[![Forkers repo roster for @devXprite/httpfy](https://reporoster.com/forks/dark/devXprite/httpfy)](https://github.com/devXprite/httpfy/network/members)


## License

Distributed under the GPL-3.0 license. See `LICENSE.txt` for more information.

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [1.0.0](https://github.com/devXprite/httpfy/compare/v0.0.7...v1.0.0) (2022-09-08)


### Documentation 

* update readme ([d3e8214](https://github.com/devXprite/httpfy/commits/d3e82140ef00e97b1f158747ece3438e4ab88159))

### [0.0.7](https://github.com/devXprite/httpfy/compare/v0.0.6...v0.0.7) (2022-08-31)


### Features

* add output folder ([ad892dd](https://github.com/devXprite/httpfy/commits/ad892ddcc008a16ae21a09c9b99b79594ac6ae49))


### Documentation 

* **readme:** update badges ([a104e33](https://github.com/devXprite/httpfy/commits/a104e331c72882a3960625a60b9bb1b8e65409e4))
* **readme:** update preview ([fe75c76](https://github.com/devXprite/httpfy/commits/fe75c7648252b5f167ae864c2bdbba2471825b07))
* update preview ([266f714](https://github.com/devXprite/httpfy/commits/266f714fe02892bae35ef1663c4e653900ffb500))

### [0.0.6](https://github.com/devXprite/httpfy/compare/v0.0.5...v0.0.6) (2022-08-18)


### Features

* support custom header ([92a1692](https://github.com/devXprite/httpfy/commits/92a1692d83fbcef57ded76f2aa06325bd89d5b0b))


### Documentation 

* Create FUNDING.yml ([e9d115d](https://github.com/devXprite/httpfy/commits/e9d115d05e88ac925d5caa4a157b30f1fd42f953))
* **readme:** add examples ([91292c2](https://github.com/devXprite/httpfy/commits/91292c22ac04175fc36315a558971c2ea62c8c40))
* **readme:** add Usages ([c6c5aa1](https://github.com/devXprite/httpfy/commits/c6c5aa1eb42a14b95e3652fccf27397ddc4bf64b))

### [0.0.5](https://github.com/devXprite/httpfy/compare/v0.0.4...v0.0.5) (2022-08-17)


### Performance

* add .versionrc and update postinstall script ([5c09bc9](https://github.com/devXprite/httpfy/commits/5c09bc940c7604684d1c0121f818a9781d9c066e))

### [0.0.4](https://github.com/devXprite/httpfy/compare/v0.0.3...v0.0.4) (2022-08-17)


### Features

* add PostInstall Script ([753ad43](https://github.com/devXprite/httpfy/commits/753ad43945a12a52ad4dabf56544d06273cb89bd))


lib/args.js
#!/usr/bin/env node

/* eslint-disable unicorn/prevent-abbreviations */
/* eslint-disable max-len */
const { Command } = require("commander");
const pkg = require("../package.json");
const banner = require("./banner");

const program = new Command();

program.name(pkg.name).description(pkg.description).version(pkg.version, "-v, --version", "output the current version");

program.requiredOption("-f, --file <path>", "Input file containing list of URLs");

program
    .option("-sc, --status-code", "display response status-code")
    .option("-cl, --content-length", "display response content-length")
    .option("-ct, --content-type", "display response content-type")
    .option("-rt, --response-time", "display response time")
    .option("-lc, --line-count", "display response body line count")
    .option("-wc, --word-count", "display response body word count")
    .option("-ws, --web-serve", "display web server name")
    .option("-rl, --redirect-location", "display redirect location")
    .option("-m, --method", "display http request method")
    .option("-nc, --no-color", "disable colors in cli output")
    .option("-ttl, --title", "display page title")
    .option("-fl, --failed", "display failed request's");

program
    .option("-nr, --no-redirect", "don't follow redirects")
    .option("-mr, --max-redirect <number>", "maximum redirects to follow")
    .option("-t, --threads <number>", "maximum cocurrent requests send", 100)
    .option("-i, --interval <number>", "interval between each thread")
    .option("-c, --cookie <string>", "send cookies (--cookie 'login=Yes')")
    .option("-x, --request-methods <string>", "set request methods, use 'all' to probe all HTTP methods")
    .option("-H, --header <string>", "custom http headers to send (--header 'X-MyHeader: XYZ')")
    .option("-pc, --protocol <string>", "set request Protocol", "https")
    .option("-path, --request-path <string>", "path or list of paths  (-path admin,login)")
    .option("-param, --request-param <string>", "set request parameters  (-param id=5)")
    .option("-ua, --user-agent <string>", "set custom useragent")
    .option("-time, --timeout <number>", "set request timeout in seconds");

program
    .option("-mc, --match-code <string>", "match response with specified status code (-mc 200,302)")
    .option("-ml, --match-length <string>", "match response with specified content length (-ml 100,102)")
    .option("-mlc, --match-line-count <string> ", "match response body with specified line count (-mlc 423,532)")
    .option("-ms, --match-string <string> ", "match response with specified strings  (-ms admin)");

program
    .option("-o , --output-file <string>", "save results in a single file  (-o result.txt)")
    .option("-of, --output-folder <string>", "save results in multiple files accoring status code  (-of result)");

program.action((args) => {
    console.log(banner(args.color));
});

program.parse();

// console.log(program.opts());

module.exports = program.opts();
